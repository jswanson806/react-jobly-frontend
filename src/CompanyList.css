.Company-list {
    margin: 0 auto;
    max-width: 600px;
    padding: 40px;
    background-color: #F5E7E1;
    border-radius: 10px;
    box-shadow: 0 8px 20px rgba(0, 0, 0, 0.15);
    font-family: 'Raleway', sans-serif;
  }
  
  .Company-list-title {
    text-align: center;
    margin-bottom: 1.5rem;
  }
  
  .Company-list-heading {
    font-size: 3rem;
    color: #618985;
  }

  .Company-list-search-form {
    display: flex;
    justify-content: flex-end;
  }
  
  .Company-list-search-label {
    font-size: 1.2rem;
    margin-top: .25em;
    margin-bottom: .5rem;
    margin-right: .25em;
    color: #475261;
  }
  
  .Company-list-search-input {
    flex: 1;
    padding: 8px;
    border-radius: 5px;
    border: 1px solid #E6D1C9;
    margin-bottom: 1rem;
    font-size: 1rem;
  }
  
  .Company-list-search-button {
    background-color: #FF725C;
    color: #fff;
    padding: 8px 16px;
    border: none;
    border-radius: 5px;
    margin-top: .1em;
    margin-bottom: auto;
    margin-left: 1em;
    align-self: center;
    transition: background-color 0.3s ease;
  }
  
  .Company-list-search-button:hover {
    background-color: #FF6B5B;
    cursor: pointer;
  }

  .Company-list-cards a {
    display: flex;
    flex-direction: column;
    margin-bottom: 4em;
  }