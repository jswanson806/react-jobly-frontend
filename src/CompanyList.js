import React, { useEffect, useState } from "react";
import CompanyCard from "./CompanyCard.js";
import { Link } from "react-router-dom";
import './styles/CompanyList.css';

/**
 * CompanyList Component
 *
 * Displays a list of companies and provides a search functionality to filter companies.
 *
 * Props:
 * - companies: An array of company objects.
 * - filterItems: A function to filter companies based on search filters.
 *
 * State:
 * - companiesList: Tracks the state of the filtered companies.
 * - companyCards: Tracks the state of the rendered CompanyCard components.
 * - formData: Tracks the state of the search form data.
 *
 * Hooks:
 * - useEffect: Maps the companies in companiesList to CompanyCard components when companiesList or companies prop changes.
 *
 * Functions:
 * - handleChange: Updates the formData state as the user types in the search input.
 * - handleSubmit: Submits the search form, calls the filterItems function, and updates the companiesList state.
 */

const CompanyList = ({companies, filterItems}) => {
    // initial value of formData
    const INITIAL_STATE = {
        name: ""
    }

    // ********************* State *********************

    // track state of companiesList -> default value is all companies
    const [companiesList, setCompaniesList] = useState(companies);
    // track state of companyCards
    const [companyCards, setCompanyCards] = useState([]);
    const [formData, setFormData] = useState(INITIAL_STATE)

    // ********************* Hooks *********************

    // hook that maps the companies in companiesList to CompanyCard components
    useEffect(() => {
        setCompanyCards(companiesList.map(company => (
            <div className="Company-list-cards" key={company.handle}>
                <Link to={`/companies/${company.handle}`} data-testid={`company-link-${company.handle}`}>
                    <CompanyCard company={company}/>
                </Link>
            </div>
        )));
        // watch for state of companiesList to update before running again
    }, [companiesList])

    // ********************* Functions *********************

    // update the form data as the user types in the input
    const handleChange = (e) => {
        const { name, value } = e.target;
        // update formData state with input field name and value of input field
        setFormData(formData => ({
          ...formData,
          [name]: value,
        }));
      }

    // submit the form data 
    const handleSubmit = async (e) => {
        e.preventDefault();
        // set the input as the value of 'name' in filters
        const filters = {
            name: formData.name
        }
        // pass filters object to function call 'filterCompanies' from parent component 'Routes.js'
        const res = await filterItems(filters);
        // update companiesList state -> triggers useEffect hook to map new companyCards
        setCompaniesList(res.companies)
        // reset search input
        setFormData(INITIAL_STATE);
    }

    return (
        <div className="Company-list" data-testid="Company-list">
            <div className="Company-list-title" data-testid="Company-list-title">
                <h1 className="Company-list-heading" data-testid="Company-list-heading">Companies</h1>
            </div>
            <div data-testid="Company-list-search">
                <form className="Company-list-search-form" onSubmit={handleSubmit} data-testid="Company-list-search-form">
                    <label htmlFor="compSearch" className="Company-list-search-label">Search</label>
                    <input type="text" 
                            id="compSearch" 
                            name="name" 
                            data-testid="Company-list-search-input"
                            value={formData.name}
                            onChange={handleChange}
                            className="Company-list-search-input"
                    />
                    <button type="submit" 
                            data-testid="Company-list-search-button"
                            className="Company-list-search-button"
                    >
                        Search
                    </button>
                </form>
            </div>
            {companyCards}
        </div>
    );
};

export default CompanyList;