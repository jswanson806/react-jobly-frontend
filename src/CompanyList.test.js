import React from 'react';
import { MemoryRouter} from 'react-router-dom';
import { act, render, waitFor, screen, fireEvent} from '@testing-library/react';
import CompanyList from './CompanyList.js';
import '@testing-library/jest-dom';


const companies = [
    {
      handle: "test-company-1",
      name: "Test Company 1",
      num_employees: 9999,
      description: "Test description for Test Company 1",
    },
    {
      handle: "test-company-2",
      name: "Test Company 2",
      num_employees: 99999,
      description: "Test description for Test Company 2",
    }
];



describe('CompanyList smoke and snapshot', () => {

    it('should render without crashing', async () => {
        await act(async () => {render(

                <MemoryRouter>
                    <CompanyList companies={companies} />
                </MemoryRouter>

            );
        });
    });

    it('should match the snapshot', async () => {
        let asFragment

        await act(async () => {const {asFragment: fragment} = render(

                <MemoryRouter>
                    <CompanyList companies={companies} />
                </MemoryRouter>

            );
            asFragment = fragment;
        });
        expect(asFragment()).toMatchSnapshot();
    });

});

describe("Renders correct content",() => {

    it('should render company cards', async () => {
        await act(async () => {render(

                <MemoryRouter>
                    <CompanyList companies={companies} />
                </MemoryRouter>

            ); 
        });

        // wait for the content to finish rendering
        await waitFor(() => {
            expect(screen.getAllByTestId("Company-description").length).toBeGreaterThan(0);
        })
        // rendered 2 company cards because companies contains only 2 companies
        expect(screen.getAllByTestId("Company-description").length).toBe(2)
    })


    it('should render text field with empty string value', async () => {
        await act(async () => {render(

                <MemoryRouter>
                    <CompanyList companies={companies} />
                </MemoryRouter>

            ); 
        });

        expect(screen.getByTestId("Company-list-search-input")).toHaveValue("");
    })

})

describe("Handles form input correctly", () => {


    it('updates search input correctly', async () => {
        await act(async () => {render(

                <MemoryRouter>
                    <CompanyList companies={companies} />
                </MemoryRouter>

            ); 
        });


      const searchInput = screen.getByTestId("Company-list-search-input");
      fireEvent.change(searchInput, {target: {value: "1"}});
      expect(searchInput).toHaveValue("1");

    })

})


describe("CompanyList", () => {
    it("should handle form submission", async () => {
      // Mock the filterItems function from the parent component
      const mockFilterItems = jest.fn().mockResolvedValue({ companies: []});

      const mockCompanies = [
        { handle: "company1" },
        { handle: "company2" },
        { handle: "company3" },
      ];
  
      // Render the component
      await act(async () => {render(
        <MemoryRouter>
            <CompanyList companies={mockCompanies} filterItems={mockFilterItems} />
        </MemoryRouter>
      );

      })
  
        // Get the search input and submit button
        const searchInput = screen.getByLabelText("Search");
        const submitButton = screen.getByRole("button", { name: "Search" });

        // Enter a search query
        fireEvent.change(searchInput, { target: { value: "test" } });

        await act(async () => {
            // Submit the form
            fireEvent.click(submitButton);
        })
        

        // Expect the filterItems function to be called with the correct filters
        expect(mockFilterItems).toHaveBeenCalledWith({ name: "test" });

    });

  });