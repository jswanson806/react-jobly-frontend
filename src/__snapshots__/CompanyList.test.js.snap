// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyList smoke and snapshot should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Company-list"
    data-testid="Company-list"
  >
    <div
      class="Company-list-title"
      data-testid="Company-list-title"
    >
      <h1
        class="Company-list-heading"
        data-testid="Company-list-heading"
      >
        Companies
      </h1>
    </div>
    <div
      data-testid="Company-list-search"
    >
      <form
        class="Company-list-search-form"
        data-testid="Company-list-search-form"
      >
        <label
          class="Company-list-search-label"
          for="compSearch"
        >
          Search
        </label>
        <input
          class="Company-list-search-input"
          data-testid="Company-list-search-input"
          id="compSearch"
          name="name"
          type="text"
          value=""
        />
        <button
          class="Company-list-search-button"
          data-testid="Company-list-search-button"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="Company-list-cards"
    >
      <a
        data-testid="company-link-test-company-1"
        href="/companies/test-company-1"
      >
        <div
          class="Company-card"
          data-testid="Company-card"
        >
          <div
            class="Company-heading"
            data-testid="Company-heading"
          >
            <h4
              class="Company-title"
              data-testid="Company-title"
            >
              Test Company 1
            </h4>
          </div>
          <div
            class="Company-description"
            data-testid="Company-description"
          >
            Test description for Test Company 1
          </div>
        </div>
      </a>
    </div>
    <div
      class="Company-list-cards"
    >
      <a
        data-testid="company-link-test-company-2"
        href="/companies/test-company-2"
      >
        <div
          class="Company-card"
          data-testid="Company-card"
        >
          <div
            class="Company-heading"
            data-testid="Company-heading"
          >
            <h4
              class="Company-title"
              data-testid="Company-title"
            >
              Test Company 2
            </h4>
          </div>
          <div
            class="Company-description"
            data-testid="Company-description"
          >
            Test description for Test Company 2
          </div>
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
